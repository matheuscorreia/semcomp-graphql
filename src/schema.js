/* @flow */
import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString,
  GraphQLBoolean, GraphQLNonNull, GraphQLID,
  GraphQLList,
} from 'graphql';
import User from './model/User';

import MutationType from './MutationType';

const UserType = new GraphQLObjectType({
  name: 'User',
  description: 'an user',
  fields: () => ({
    _id: {
      type: GraphQLString,
      resolve: user => user._id,
    },
    name: {
      type: GraphQLString,
      resolve: user => user.name,
    },
    email: {
      type: GraphQLString,
      resolve: (user, args, context) => {
        user.email
      },
    },
    active: {
      type: GraphQLBoolean,
      resolve: user => user.active,
    },
  }),
});

const QueryType = new GraphQLObjectType({
  name: 'Query',
  description: 'The root of all... queries',
  fields: () => ({
    hello: {
      type: GraphQLString,
      args: {
        name: {
          type: GraphQLString,
          description: 'person name',
        },
      },
      description: 'my first program on graphql',
      resolve: (root, args, context) => args.name ? `hello ${args.name}` : 'hello stranger',
    },
    user: {
      type: UserType,
      args: {
        id: {
          type: new GraphQLNonNull(GraphQLID),
        },
      },
      resolve: (root, args, context) => {
        return User.findOne({
          _id: args.id,
        });
      },
    },
    users: {
      type: new GraphQLList(UserType),
      resolve: async (_,  args, context) => {
        return User.find({});
      },
    },
  }),
});

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});
