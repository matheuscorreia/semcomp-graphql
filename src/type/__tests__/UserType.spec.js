/* @flow */
import { graphql } from 'graphql';
import { schema } from '../../schema';
import {
  User,
  Conversation,
} from '../../model';
import {
  getContext,
  setupTest,
  sanitizeTestObject,
} from '../../../test/helper';

beforeEach(async () => await setupTest());

it('should get an user', async () => {
  const user = new User({
    name: 'user',
    email: 'user@example.com',
    password: '123',
  });
  await user.save();

  //language=GraphQL
  const query = `
    query Q {
      user(id: "${user._id.toString()}") {
        _id
        name
      }
    }
  `;

  const rootValue = {};
  const context = getContext({ user });

  const result = await graphql(schema, query, rootValue, context);

  expect(sanitizeTestObject(result)).toMatchSnapshot();
});

it('should get an user', async () => {
  const user = new User({
    name: 'user',
    email: 'user@example.com',
    password: '123',
  });
  await user.save();

  const conversation = new Conversation({
    name: 'ok',
    users: [
      user,
    ],
    owner: user,
    active: true,
  });
  await conversation.save();

  //language=GraphQL
  const query = `
    query Q {
      user(id: "${user._id.toString()}") {
        _id
        name
        conversations {
          _id
          name
        }
      }
    }
  `;

  const rootValue = {};
  const context = getContext({ user });

  const result = await graphql(schema, query, rootValue, context);

  expect(sanitizeTestObject(result)).toMatchSnapshot();
});


