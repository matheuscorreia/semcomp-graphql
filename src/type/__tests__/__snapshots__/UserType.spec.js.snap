// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get an user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "_id": "ObjectId",
      "name": "user",
    },
  },
}
`;

exports[`should get an user 2`] = `
Object {
  "data": Object {
    "user": Object {
      "_id": "ObjectId",
      "conversations": Array [
        Object {
          "_id": "ObjectId",
          "name": "ok",
        },
      ],
      "name": "user",
    },
  },
}
`;
